napari/layers/surface/_tests/test_surface.py

```diff
@@ -493,6 +493,17 @@ def test_surface_copy():
     assert l1.data[0] is not l2.data[0]
 
 
+def test_surface_with_no_visible_faces():
```

Comment:
This test is not failing

Reply:
I made a small script to test this, I didn't test it in the test environment. 

```
import numpy as np
from napari.layers import Surface
from napari._vispy.layers.surface import VispySurfaceLayer
import napari

points = np.array(
    [
        [0, 0.0, 0.0, 0.0],
        [0, 1., 0, 0],
        [0, 1, 1, 0],
        [2, 0.0, 0.0, 0.0],
        [2, 1., 0, 0],
        [2, 1, 1, 0]
        ] );
faces = np.array([[0, 1, 2], [3, 4, 5]])
layer = Surface((points, faces))
viewer = napari.Viewer(ndisplay=3)
viewer.add_layer(layer)
viewer.reset()
#layer._set_view_slice()
#vl = VispySurfaceLayer(layer)
#layer._set_view_slice()
#vl._on_data_change()
napari.run()

```
I use checkout 9fcf63e69ac61b5dff0259c8618f828cc5169a9c 


This is the error I get:

```
Traceback (most recent call last):
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/app/backends/_qt.py", line 928, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/util/event.py", line 453, in __call__
    self._invoke_callback(cb, event)
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/util/event.py", line 471, in _invoke_callback
    _handle_exception(self.ignore_callback_errors,
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/util/event.py", line 469, in _invoke_callback
    cb(event)
  File "/home/username/linux-desktop/napari/napari-repo/napari/_vispy/camera.py", line 188, in on_draw
    self._camera.angles = self.angles
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/evented_model.py", line 307, in __setattr__
    with ComparisonDelayer(self):
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/evented_model.py", line 520, in __exit__
    self._target._check_if_values_changed_and_emit_if_needed()
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/evented_model.py", line 345, in _check_if_values_changed_and_emit_if_needed
    getattr(self.events, name)(value=new_value)
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/event.py", line 764, in __call__
    self._invoke_callback(cb, event if pass_event else None)
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/event.py", line 802, in _invoke_callback
    _handle_exception(
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/event.py", line 789, in _invoke_callback
    cb(event)
  File "/home/username/linux-desktop/napari/napari-repo/napari/_vispy/layers/surface.py", line 208, in _on_camera_move
    self.node.shading_filter.light_dir = self._light_direction
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/visuals/filters/mesh.py", line 460, in light_dir
    self._update_data()
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/visuals/filters/mesh.py", line 552, in _update_data
    normals = self._visual.mesh_data.get_vertex_normals(indexed='faces')
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/geometry/meshdata.py", line 380, in get_vertex_normals
    face_normals = self.get_face_normals()
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/geometry/meshdata.py", line 351, in get_face_normals
    self._face_normals = _compute_face_normals(vertices)
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/geometry/meshdata.py", line 23, in _compute_face_normals
    if vertices.shape[1:] != (3, 3):
AttributeError: 'NoneType' object has no attribute 'shape'
Traceback (most recent call last):
  File "/home/username/linux-desktop/napari/geometry_fail.py", line 19, in <module>
    viewer.reset()
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/evented_model.py", line 412, in reset
    getattr(self, name).reset()
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/evented_model.py", line 417, in reset
    setattr(self, name, value)
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/evented_model.py", line 307, in __setattr__
    with ComparisonDelayer(self):
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/evented_model.py", line 520, in __exit__
    self._target._check_if_values_changed_and_emit_if_needed()
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/evented_model.py", line 345, in _check_if_values_changed_and_emit_if_needed
    getattr(self.events, name)(value=new_value)
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/event.py", line 764, in __call__
    self._invoke_callback(cb, event if pass_event else None)
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/event.py", line 802, in _invoke_callback
    _handle_exception(
  File "/home/username/linux-desktop/napari/napari-repo/napari/utils/events/event.py", line 789, in _invoke_callback
    cb(event)
  File "/home/username/linux-desktop/napari/napari-repo/napari/_vispy/layers/surface.py", line 208, in _on_camera_move
    self.node.shading_filter.light_dir = self._light_direction
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/visuals/filters/mesh.py", line 460, in light_dir
    self._update_data()
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/visuals/filters/mesh.py", line 552, in _update_data
    normals = self._visual.mesh_data.get_vertex_normals(indexed='faces')
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/geometry/meshdata.py", line 380, in get_vertex_normals
    face_normals = self.get_face_normals()
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/geometry/meshdata.py", line 351, in get_face_normals
    self._face_normals = _compute_face_normals(vertices)
  File "/home/username/linux-desktop/napari/napari-venv/lib/python3.10/site-packages/vispy/geometry/meshdata.py", line 23, in _compute_face_normals
    if vertices.shape[1:] != (3, 3):
AttributeError: 'NoneType' object has no attribute 'shape'

```

Maybe it has been fixed. In the latest version of main?

Reply:
ah. You make pr from your main branch. 
It looks like this test requires some additional steps to execute the event loop.

Reply:
Yes, I hadn't done a fork/pull request, so I probably should have done a separate branch. I also should have probably made 2 pull requests. The integer one seems to be a good practice. 

For the VispyLayer error. I didn't see where or how to correctly create the vispy layer and have it update the data with the shader attached. It also seems to be a regression. When I used napari from pip I didn't get the error. Once I pulled the latest from main I was getting the error.

Reply:
Ok. I will try to do this today. Did you understand my first suggestions (to other test)?

Reply:
Actually, I am a bit confused now. Are you refering to this change? `with pytest.raises(ValueError, match="operands could not")`

Reply:
yes. This is a valid way to check if expected exception was raised. We provide exception class and part of text to ensure. Without this, we have failing test run

Reply:
My bad. I switch tests when reading. I have pushed changes to make test working

Rename from `check_` to `test_`. Use `make_napari_viewer` instead of direct call `Viever` and show it.  Add `with pytest.raises(...)`